### PR TITLE
fix(proxy) give manager leader control over start

### LINKED ISSUE #2052: Leader election defaults and implementation do not result in expected behavior
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

Originally reported internally by FTI-3034.

Leader election is disabled by default. When using database-backed instances, all controllers attempt to send config updates and may cause config thrashing.

When leader election is enabled, it disables all controllers but not the sync or status loops. Non-leader controllers update config incorrectly, thinking they need to clear out all configuration because there are no resources loaded into the cache. Probably nothing happens with status because AFAIK it won't check for anything that doesn't actually result in a Kong route.

Furthermore, you can enable leader election in DB-less mode, which just results in many active, but unusable (no config) proxy instances.

### Expected Behavior

When using DB-less mode, all instances act as leaders, and push configuration to their proxy instance.

When using a database, a single leader updates configuration. Other instances attempt to become the leader if the current leader disappears, but otherwise do nothing other than report readiness/liveness/metrics info.

Likely do not expose leader election configuration to users at all. Off is always appropriate for DB-less, on is always appropriate for database-backed.

### Steps To Reproduce

```markdown
Manifests in many ways, but IMO the simplest example is start a Postgres-backed instance with election on, scale to two, and watch logs of the second pod. It should not be a leader, but will interact with entities via the admin API, and will delete entities it shouldn't.
```

### Kong Ingress Controller version

```shell
2.0.x
```

### Kubernetes version

_No response_

### Anything else?

In 1.x, leader election was a bespoke implementation and was always enabled, but DB-less largely ignored it and syncs anyway: https://github.com/Kong/kubernetes-ingress-controller/blob/v1.3.3/internal/ingress/controller/controller.go#L96-L100

In 2.x, leader election is a feature of [the controller-runtime manager](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/manager). Observation indicates it uses this in several places internally to decide if a task should execute, e.g. the readiness and liveness endpoint server does run on non-leaders, whereas controllers registered via `NewControllerManagedBy()` do not. Not all functionality is directly executed by the Manager, however: among others (the admission server, status update system), the proxy cache server (which handles sending config updates) is executed [via a goroutine](https://github.com/Kong/kubernetes-ingress-controller/blob/v2.0.6/internal/proxy/clientgo_cached_proxy_resolver.go#L90) which we set up [independent of](https://github.com/Kong/kubernetes-ingress-controller/blob/v2.0.6/internal/manager/run.go#L65) calling [manager.Start()](https://github.com/Kong/kubernetes-ingress-controller/blob/v2.0.6/internal/manager/run.go#L115).

### PR DESCRIPTION
**Pull request by rainest**
**What this PR does / why we need it**:
- Refactors the proxy update and status update subsystems into [LeaderElectionRunnables](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/manager#LeaderElectionRunnable) and starts them via `manager.Add()`. These subsystems now only run when an instance is a leader. This fixes the part of #2052 where enabling leader election would actually attempt to push a null configuration because the proxy update loop was running, but the controllers were not. While the status updater caused no issues (it will simply do nothing when given an empty configuration), running multiple instances of it was pointless.
- Removes the `--elect-leader` flag in favor of determining whether to elect a leader automatically based on the database mode (DB-less does not use leaders, database-backed does). This is a minor breaking change (the controller will reject the removed flag if present). We could optionally leave it as a deprecated no-op flag if we don't want to make the breaking change immediately. 
- Adds an `--election-namespace` flag. This is necessary to run integration tests when leader election is enabled, as the controller has no way of determining the election namespace when run outside a cluster. It is unlikely to have any production use.
- Adds a new multi-replica Postgres controller test. The test confirms leadership behavior based on whether metrics report that the controller has successfully synced configuration.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #2052 

**Special notes for your reviewer**:
- Backporting this is iffy, since this fails tests without the new `--election-namespace` flag. We could remove that and the automatic mode detection, but in that case we simply don't test leader election mode in the backport. This is maybe okay since the flag is only needed for test instances (unless you run leader-election-enabled instances outside the cluster in prod for some reason, but you can't do that with current 2.0.x anyway).
- This is a new E2E test that fails against the current version. An example override run versus 2.0.x is available in [postgres_replicas.txt](https://github.com/Kong/kubernetes-ingress-controller/files/7715063/postgres_replicas.txt). Merging this will cause nightly tests to fail that test. We already have one such test from https://github.com/Kong/kubernetes-ingress-controller/pull/2005. We should decide how we want to handle these better, probably by running nightly E2E against nightly builds.
- The admission controller behavior is an open question. We make use of the [Kubernetes client to load other resources](https://github.com/Kong/kubernetes-ingress-controller/blob/994fa0404344e97a014366789a3f349dfd3d812d/internal/admission/utils.go#L50-L53) for the consumer/credential checks. I'm unsure how these behave on non-leaders. This PR shouldn't change anything about that, since it's only disabling proxy updates in addition to the controllers, which were already disabled. However, since leader election was previously a manual flag (and was not in 1.x), it's quite possible that nobody had enabled it to observe any problem behavior. I don't think we can simply disable it because the Service used in the webhook configuration points at the entire Deployment. 

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [x] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

### CODE REVIEW

File: test/e2e/all_in_one_test.go [hunk 1]
```diff
@@ -230,6 +236,115 @@ func TestDeployAllInOnePostgres(t *testing.T) {
...
+	verifyIngress(ctx, t, env)
+
+	forDeployment := metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("app=%s", deployment.Name),
+	}
+
+	podList, err := env.Cluster().Client().CoreV1().Pods(deployment.Namespace).List(ctx, forDeployment)
+	require.NoError(t, err)
+	require.Equal(t, 1, len(podList.Items))
+	initialPod := podList.Items[0]
```
Comment (shaneutt) [inline test/e2e/all_in_one_test.go:281]:
```suggestion

	t.Log("verifying that kong pods deployed properly and gathering a sample pod")
	forDeployment := metav1.ListOptions{
		LabelSelector: fmt.Sprintf("app=%s", deployment.Name),
	}
	podList, err := env.Cluster().Client().CoreV1().Pods(deployment.Namespace).List(ctx, forDeployment)
	require.NoError(t, err)
	require.Equal(t, 1, len(podList.Items))
	initialPod := podList.Items[0]
```

File: test/e2e/all_in_one_test.go [hunk 2]
```diff
@@ -230,6 +236,115 @@ func TestDeployAllInOnePostgres(t *testing.T) {
...
+			Name:      deployment.Name,
+			Namespace: deployment.Namespace,
+		},
+		Spec: autoscalingv1.ScaleSpec{
+			Replicas: 2,
+		},
+	}
+	t.Log("adding a second replica to the Kong deployment")
+	_, err = env.Cluster().Client().AppsV1().Deployments(deployment.Namespace).UpdateScale(ctx,
+		deployment.Name, scale, metav1.UpdateOptions{})
```
Comment (shaneutt) [inline test/e2e/all_in_one_test.go:294]:
```suggestion
	t.Log("adding a second replica to the Kong deployment")
	scale := &autoscalingv1.Scale{
		ObjectMeta: metav1.ObjectMeta{
			Name:      deployment.Name,
			Namespace: deployment.Namespace,
		},
		Spec: autoscalingv1.ScaleSpec{
			Replicas: 2,
		},
	}
	_, err = env.Cluster().Client().AppsV1().Deployments(deployment.Namespace).UpdateScale(ctx,
		deployment.Name, scale, metav1.UpdateOptions{})
```

File: test/e2e/all_in_one_test.go [hunk 3]
```diff
@@ -230,6 +236,115 @@ func TestDeployAllInOnePostgres(t *testing.T) {
...
+	require.NoError(t, err)
+
+	require.Eventually(t, func() bool {
+		deployment, err = env.Cluster().Client().AppsV1().Deployments(deployment.Namespace).Get(ctx, deployment.Name, metav1.GetOptions{})
+		require.NoError(t, err)
+		return deployment.Status.ReadyReplicas == *deployment.Spec.Replicas
+	}, kongComponentWait, time.Second)
```
Comment (shaneutt) [inline test/e2e/all_in_one_test.go:301]:
```suggestion
	t.Log("verifying that scaling completes and the additional replicas come up")
	require.Eventually(t, func() bool {
		deployment, err = env.Cluster().Client().AppsV1().Deployments(deployment.Namespace).Get(ctx, deployment.Name, metav1.GetOptions{})
		require.NoError(t, err)
		return deployment.Status.ReadyReplicas == *deployment.Spec.Replicas
	}, kongComponentWait, time.Second)
```

File: test/e2e/all_in_one_test.go [hunk 4]
```diff
@@ -230,6 +236,115 @@ func TestDeployAllInOnePostgres(t *testing.T) {
...
+
+	podList, err = env.Cluster().Client().CoreV1().Pods(deployment.Namespace).List(ctx, forDeployment)
+	require.NoError(t, err)
+	var secondary corev1.Pod
+	for _, pod := range podList.Items {
+		if pod.Name != initialPod.Name {
+			secondary = pod
+			break
+		}
+	}
```
Comment (shaneutt) [inline test/e2e/all_in_one_test.go:311]:
```suggestion
	t.Log("gathering another sample pod to verify leadership is configured appropriately")
	podList, err = env.Cluster().Client().CoreV1().Pods(deployment.Namespace).List(ctx, forDeployment)
	require.NoError(t, err)
	var secondary corev1.Pod
	for _, pod := range podList.Items {
		if pod.Name != initialPod.Name {
			secondary = pod
			break
		}
	}
```

File: test/e2e/all_in_one_test.go [hunk 5]
```diff
@@ -230,6 +236,115 @@ func TestDeployAllInOnePostgres(t *testing.T) {
...
+
+	t.Log("confirming the second replica is not the leader and is not pushing configuration")
+	forwardCtx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	startPortForwarder(forwardCtx, t, env, secondary, "9777", "cmetrics")
```
Comment (shaneutt) [inline test/e2e/all_in_one_test.go:316]: is there a reason we're relying on a port-forward here? commonly in our tests we create a `LoadBalancer` service on kind clusters for this kind of thing and we already have test tooling relevant to this :thinking:

File: test/e2e/all_in_one_test.go [hunk 6]
```diff
@@ -230,6 +236,115 @@ func TestDeployAllInOnePostgres(t *testing.T) {
...
+	require.Never(t, func() bool {
+		req, err := http.NewRequest("GET", "http://localhost:9777/metrics", nil)
+		require.NoError(t, err)
+		client := &http.Client{}
```
Comment (shaneutt) [inline test/e2e/all_in_one_test.go:320]: As configured I believe the empty timeout here could theoretically result in a timeout that would make the function take longer than the wait time requested. In `test/integration` we provide a global client (named `httpc`) with a standard timeout defined which is no unbounded, and also shorter than the default client provided by `net/http`. I suggest we adopt something similar here, or at least add an explicit timeout to avoid weird lag issues on line 321.

File: test/e2e/all_in_one_test.go [hunk 7]
```diff
@@ -230,6 +236,115 @@ func TestDeployAllInOnePostgres(t *testing.T) {
...
+		resp, err := client.Do(req)
+		require.NoError(t, err)
+		defer resp.Body.Close()
+		body, err := io.ReadAll(resp.Body)
+		require.NoError(t, err)
+		// if we are not the leader, we run no config pushes, and this metric string will not appear
+		return strings.Contains(string(body), metrics.MetricNameConfigPushCount)
```
Comment (shaneutt) [inline test/e2e/all_in_one_test.go:327]: :+1:

File: test/e2e/all_in_one_test.go [hunk 8]
```diff
@@ -230,6 +236,115 @@ func TestDeployAllInOnePostgres(t *testing.T) {
...
+	}, time.Minute, time.Second*10)
```
Comment (shaneutt) [inline test/e2e/all_in_one_test.go:328]: It seems you could just as reasonably run this every 1 second, especially if you make your http client timeout something like 10 seconds, because this will just add unnecessary lag to the test case from what I can see.
```suggestion
	}, time.Minute, time.Second)
```

File: test/e2e/all_in_one_test.go [hunk 9]
```diff
@@ -230,6 +236,115 @@ func TestDeployAllInOnePostgres(t *testing.T) {
...
+
+	t.Log("deleting the original replica and current leader")
+	err = env.Cluster().Client().CoreV1().Pods(initialPod.Namespace).Delete(ctx, initialPod.Name, metav1.DeleteOptions{})
+	require.NoError(t, err)
+
+	t.Log("confirming the second replica becomes the leader and starts pushing configuration")
+	require.Eventually(t, func() bool {
+		req, err := http.NewRequest("GET", "http://localhost:9777/metrics", nil)
+		require.NoError(t, err)
+		client := &http.Client{}
```
Comment (shaneutt) [inline test/e2e/all_in_one_test.go:338]: Since we're also using a `net/http` client here, probably should just make a default `httpc` as suggested before and use it across both.

File: test/e2e/all_in_one_test.go [hunk 10]
```diff
@@ -230,6 +236,115 @@ func TestDeployAllInOnePostgres(t *testing.T) {
...
+		resp, err := client.Do(req)
+		require.NoError(t, err)
+		defer resp.Body.Close()
+		body, err := io.ReadAll(resp.Body)
+		require.NoError(t, err)
+		return strings.Contains(string(body), metrics.MetricNameConfigPushCount)
+	}, time.Minute, time.Second*10)
```
Comment (shaneutt) [inline test/e2e/all_in_one_test.go:345]:
```suggestion
	}, time.Minute, time.Second)
```

File: test/e2e/all_in_one_test.go [hunk 11]
```diff
@@ -508,6 +625,35 @@ func verifyPostgres(ctx context.Context, t *testing.T, env environments.Environm
...
+	require.Eventually(t, func() bool {
+		conn, err := net.Dial("tcp", fmt.Sprintf("localhost:%s", localPort))
+		if err == nil {
+			conn.Close()
+			return true
+		}
+		return false
+	}, kongComponentWait, time.Second)
+}
+
```
Comment (shaneutt) [inline test/e2e/all_in_one_test.go:656]: related to the above, we already have some tooling present for adding `LoadBalancer` services to support this kind of thing, perhaps we should continue to rely on that instead?

**Review by shaneutt — DISMISSED** [review]
Looking great! I have a few suggestions and comments I would like to see resolved before we merge but mostly LGTM.

File: CHANGELOG.md [hunk 12]
```diff
@@ -49,6 +49,12 @@ you should edit it (`kubectl edit validatingwebhookconfiguration
 kong-validations`) and add `kongclusterplugins` under the `resources` block for
 the `configuration.konghq.com` API group.
 
+#### Breaking Changes
+
+- The `--leader-elect` flag has been removed. Leader election is now enabled or
+  disabled automatically based on the database mode.
+  [#2053](https://github.com/Kong/kubernetes-ingress-controller/issues/2053)
+
```
[the remark this answers is not in the document]
Comment (rainest) [reply CHANGELOG.md:57]: Do we want the same for 2.1.x? The commit is separate for that reason, so that we can cherry-pick only parts of this into 2.0.x if desired. It may make sense to put that in a separate PR if we want it gone in 2.1 just to simplify rebase/squash/whatever at the end of this one.

Simply not removing the flag in 2.0.x would still be breaking since we no longer honor the flag, but I could rework that to separate the logic that sets the mode automatically, so that for 2.0.x you need to enable leader election manually but it doesn't delete configuration if you do.

File: internal/ctrlutils/ingress-status.go [hunk 13]
```diff
@@ -52,6 +52,48 @@ type statusInfo struct {
 	hostname string
 }
 
+//  StatusUpdater is a Kubebuilder controller that updates resource statuses
```
[the remark this answers is not in the document]
Comment (rainest) [reply internal/ctrlutils/ingress-status.go]: Part of 41c61b31

File: internal/ctrlutils/ingress-status.go [hunk 14]
```diff
@@ -52,6 +52,48 @@ type statusInfo struct {
...
+	publishAddresses []string
+}
+
+func (s *StandardStatusUpdater) NeedLeaderElection() bool {
+	return true
+}
+
+func (s *StandardStatusUpdater) Start(ctx context.Context) error {
+	go s.PullConfigUpdate(ctx)
+
```
[the remark this answers is not in the document]
Comment (rainest) [reply internal/ctrlutils/ingress-status.go]: Part of 41c61b31

File: internal/ctrlutils/ingress-status.go [hunk 15]
```diff
@@ -52,6 +52,48 @@ type statusInfo struct {
...
+type StatusUpdater interface {
+	// NeedLeaderElection indicates if the proxy must only be run when it is controlled by a leader controller
+	NeedLeaderElection() bool
+
+	// Start starts the proxy update loop
+	Start(ctx context.Context) error
+
+	// PullConfigUpdate starts the config update routine
+	PullConfigUpdate(ctx context.Context)
+}
```
[the remark this answers is not in the document]
Comment (rainest) [reply internal/ctrlutils/ingress-status.go]: Tried to do this originally returning a non-pointer and thought:

```
ctrlutils.StandardStatusUpdater does not implement manager.Runnable (Start method has pointer receiver)
```
was because it wasn't returning an interface (since that's what NewCacheBasedProxyWithStagger does). TIL it's actually whether it's a pointer or not, and the interface version has to be a pointer.

41c61b31

File: internal/ctrlutils/ingress-status.go [hunk 16]
```diff
@@ -52,6 +52,48 @@ type statusInfo struct {
...
+	return true
+}
+
+func (s *StandardStatusUpdater) Start(ctx context.Context) error {
+	go s.PullConfigUpdate(ctx)
+
+	return nil
+}
+
+func NewStandardStatusUpdater(kongConfig sendconfig.Kong, log logr.Logger, kubeConfig *rest.Config,
```
[the remark this answers is not in the document]
Comment (rainest) [reply internal/ctrlutils/ingress-status.go]: 41c61b31

File: test/e2e/all_in_one_test.go (hunk 5 above)
Comment (rainest) [reply test/e2e/all_in_one_test.go:316]: LoadBalancer distributes requests at random. We wouldn't know if we're hitting the leader or the the follower node for any individual request.

Ideally we'd use Pod DNS instead, but we're not in the cluster, so :shrug:

File: test/e2e/all_in_one_test.go (hunk 8 above)
Comment (rainest) [reply test/e2e/all_in_one_test.go:328]: `Never()` is always going to run the full lifetime (`time.Minute`) regardless.

We're kinda stuck with that since we don't have any external signalling that'd indicate "the proxy loop would have started if it was going to by now" AFAIK: in DB mode, controllers immediately mark themselves ready in `proxy.Ready()`, since they can assume Kong is pulling current config from the DB.

File: test/e2e/all_in_one_test.go (hunk 11 above)
Comment (rainest) [reply test/e2e/all_in_one_test.go:656]: Ditto above, need to hit individual specific Pods' endpoints.

The two metrics endpoints also lack Services because they provide data specific to that Pod. Normal usage has Prometheus scrape each Pod using Pod hostnames.

File: CHANGELOG.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (shaneutt) [reply CHANGELOG.md:57]: > Do we want the same for 2.1.x? The commit is separate for that reason, so that we can cherry-pick only parts of this into 2.0.x if desired. It may make sense to put that in a separate PR if we want it gone in 2.1 just to simplify rebase/squash/whatever at the end of this one.
> Simply not removing the flag in 2.0.x would still be breaking since we no longer honor the flag, but I could rework that to separate the logic that sets the mode automatically, so that for 2.0.x you need to enable leader election manually but it doesn't delete configuration if you do.

It's a tricky one because we don't want to make a technically breaking change, but the underlining here is that we know if anyone explicitly set the flag and that setting was opposed to what the automatic configuration would have chosen it would result in a broken deployment.

In other words if there's anyone out there currently using the flag we have certainty they can only be setting it the exact way our new automation would set it, so not honoring the setting doesn't break anything because it will result in the same setting. The only thing we really need to avoid is dropping the flag entirely, because then those same theoretical users would get an upgrade failure.

For that reason I'm leaning towards disabling the functionality of the flag (because we wont be changing the actual value since it _has_ to be a certain way) but also marking the flag as deprecated and removing it in `v3` so that we're actually following semver and protecting that upgrade path.

File: internal/manager/config.go [hunk 17]
```diff
@@ -130,8 +130,8 @@ func (c *Config) FlagSet() *pflag.FlagSet {
 	// Kubernetes configurations
 	flagSet.StringVar(&c.KubeconfigPath, "kubeconfig", "", "Path to the kubeconfig file.")
 	flagSet.StringVar(&c.IngressClassName, "ingress-class", annotations.DefaultIngressClass, `Name of the ingress class to route through this controller.`)
-	flagSet.BoolVar(&c.EnableLeaderElection, "leader-elect", false, "Enable leader election for controller manager. Enabling this will ensure there is only one active controller manager.")
```
[the remark this answers is not in the document]
Comment (rainest) [reply internal/manager/config.go:133]: Related to the "Do we want the same for 2.1.x?" question above.

IMO we remove this one immediately:
- We do not intend to honor it at all. At most the deprecated flag would print a warning log if not set to the default.
- I expect that usage of this flag in the wild is low: it wasn't present in 1.x and users that upgraded were accustomed to it being handled automatically, and we didn't add documentation indicating you should enable it in DB mode AFAIK. If people _were_ using it, I expect we'd have received a report about it deleting configuration when enabled, and we didn't.
- Upgrade burden is pretty small. If you weren't using it, you do nothing. If you were, you'll see your upgrade fail and print a clear "unknown flag" error, and you run a new upgrade that removes it.

File: internal/proxy/clientgo_cached_proxy_resolver.go [hunk 18]
```diff
@@ -185,19 +192,19 @@ func (p *clientgoCachedProxyResolver) Listeners(ctx context.Context) ([]kong.Pro
 
 // startProxyUpdateServer runs a server in a background goroutine that is responsible for
 // updating the kong proxy backend at regular intervals.
-func (p *clientgoCachedProxyResolver) startProxyUpdateServer() {
+func (p *clientgoCachedProxyResolver) startProxyUpdateServer(ctx context.Context) {
 	var initialConfig sync.Once
 	for {
 		select {
-		case <-p.ctx.Done():
+		case <-ctx.Done():
```
[the remark this answers is not in the document]
Comment (rainest) [reply internal/proxy/clientgo_cached_proxy_resolver.go:194]: 9ec542eb

[the remark this answers is not in the document]
Comment (rainest) [reply internal/proxy/clientgo_cached_proxy_resolver.go:194]: 80165b46b4e4d217d58ce066054068d05ece9abb

File: internal/proxy/proxy.go [hunk 19]
```diff
@@ -69,6 +69,12 @@ type Proxy interface {
 	// underlying proxy so that callers can gather this metadata to
 	// know which ports and protocols are in use by the proxy.
 	Listeners(ctx context.Context) ([]kong.ProxyListener, []kong.StreamListener, error)
+
+	// NeedLeaderElection indicates if the proxy must only be run when it is controlled by a leader controller
+	NeedLeaderElection() bool
+
+	// Start starts the proxy update loop
+	Start(ctx context.Context) error
```
[the remark this answers is not in the document]
Comment (rainest) [reply internal/proxy/proxy.go]: 43fb6668959bf6b929e4696c142521271b7cdf7e

File: internal/manager/config.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (rainest) [reply internal/manager/config.go:133]: Resolving this one and continuing in the CHANGELOG.md thread.

[the remark this answers is not in the document]
Comment (shaneutt) [reply internal/manager/config.go:133]: > We do not intend to honor it at all. At most the deprecated flag would print a warning log if not set to the default.

True, but we also know that what we set it to is what the end-user would have to have set it to or it would have broken their deployment.

> I expect that usage of this flag in the wild is low: it wasn't present in 1.x and users that upgraded were accustomed to it being handled automatically, and we didn't add documentation indicating you should enable it in DB mode AFAIK. If people were using it, I expect we'd have received a report about it deleting configuration when enabled, and we didn't.
> Upgrade burden is pretty small. If you weren't using it, you do nothing. If you were, you'll see your upgrade fail and print a clear "unknown flag" error, and you run a new upgrade that removes it.

I get where you're coming from, but we have two threads going now so I'm going to try and stick with https://github.com/Kong/kubernetes-ingress-controller/pull/2053#discussion_r769681802 and follow up with you there.

File: CHANGELOG.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (rainest) [reply CHANGELOG.md:57]: IMO we remove this one immediately:

- We do not intend to honor it at all. At most the deprecated flag would print a warning log if not set to the default.-
- I expect that usage of this flag in the wild is low: it wasn't present in 1.x and users that upgraded were accustomed to it being handled automatically, and we didn't add documentation indicating you should enable it in DB mode AFAIK. If people were using it, I expect we'd have received a report about it deleting configuration when enabled, and we didn't.
- Upgrade burden is pretty small. If you weren't using it, you do nothing. If you were, you'll see your upgrade fail and print a clear "unknown flag" error, and you run a new upgrade that removes it.

Had it been around longer I'd be less inclined to remove it quickly, but as 2.0.x is the only version that includes it, I figure we remove it now and make it clear it's gone to anyone that set it and didn't read the changelog.

File: internal/manager/config.go [hunk 20]
```diff
@@ -49,13 +49,13 @@ type Config struct {
...
-	Concurrency          int
-	FilterTags           []string
-	WatchNamespaces      []string
+	KubeconfigPath          string
+	IngressClassName        string
+	LeaderElectionNamespace string
+	LeaderElectionID        string
+	Concurrency             int
+	FilterTags              []string
+	WatchNamespaces         []string
```
[the remark this answers is not in the document]
Comment (rainest) [reply internal/manager/config.go:58]: I'm a bit leery of placing test getenv calls in the actual manager setup/any non-test code (we currently only have them in the test code), and prefer having a single source of truth for actual manager configuration.

This isn't the first such flag that isn't used outside of strange scenarios: `–kubeconfig` and `–apiserver-host` similarly don't have an obvious usage outside of test instances outside a cluster.

File: internal/manager/run.go [hunk 21]
```diff
@@ -62,8 +61,22 @@ func Run(ctx context.Context, c *Config, diagnostic util.ConfigDumpDiagnostic) e
...
+	kongRootConfig, ok := kongRoot["configuration"].(map[string]interface{})
+	if !ok {
+		return fmt.Errorf("invalid root configuration, expected a map[string]interface{} got %T",
+			kongRoot["configuration"])
+	}
+	dbmode, ok := kongRootConfig["database"].(string)
+	if !ok {
+		return fmt.Errorf("invalid database configuration, expected a string got %T", kongRootConfig["database"])
+	}
+
```
[the remark this answers is not in the document]
Comment (rainest) [reply internal/manager/run.go:77]: Is there some way we could expose a generic "get arbitrary JSON path from the root output" (or any output, I guess)? That'd simplify this general pattern a lot, and avoid a lot of similar functions (since the root output is quite large).

I didn't DRY here since run.go is the abnormal area of code we don't have access to this already from the proxy object (sort of--we don't have a `GetDBMode()` but could add one easily)

File: test/e2e/all_in_one_test.go (hunk 6 above)
Comment (rainest) [reply test/e2e/all_in_one_test.go]: dc6a91039b7899a03ae49a8b2cda99483e0ce090

File: test/e2e/all_in_one_test.go (hunk 9 above)
Comment (rainest) [reply test/e2e/all_in_one_test.go]: dc6a91039b7899a03ae49a8b2cda99483e0ce090

File: CHANGELOG.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (shaneutt) [reply CHANGELOG.md]: We use semver as a contract with our end-users that states we wont make backwards-incompatible changes except in major release versions. We assume, but we don't actually _know_ that nobody is using this flag explicitly and for all we know there's an example floating around on the internet somewhere that people are using which has this flag in it. Semver as a precept can often be used _instead of_ such telemetry: you can't always know what to expect of your end-users, but your end-users should be able to know what to expect of you. Given that the actual value of this configuration is inevitably going to match our automation, and given how small of a burden it would be on us to add a warning and some documentation and then mark this deprecated until `v3` where we can remove the few deprecated flags we have, I don't feel persuaded towards the position you're advocating for. However given how small this does feel in the grand scheme of things, and given that my perspective on this is to my knowledge and after reviewing our documentation _not_ something that I can see we ever unanimously agreed to institutionalize I also can't justify a hard block. As such I've dismissed my block on the PR and I invite @mflendrich to weigh in.

File: internal/manager/config.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (shaneutt) [reply internal/manager/config.go:58]: Fair enough. I don't feel strongly enough about this to block it.

File: internal/manager/run.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (shaneutt) [reply internal/manager/run.go:77]: > Is there some way we could expose a generic "get arbitrary JSON path from the root output" (or any output, I guess)? That'd simplify this general pattern a lot, and avoid a lot of similar functions (since the root output is quite large).

I've used https://github.com/tidwall/gjson in the past, however using this as opposed to having JSON annotated structs in our codebases feels _a little bit wrong_ if nothing else because of the precedent being strong in our codebases for the annotation method of actually defining the structure.

File: CHANGELOG.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (mflendrich) [reply CHANGELOG.md]: I strongly agree with @shaneutt's view quoted below:

> Given that the actual value of this configuration is inevitably going to match our automation, and given how small of a burden it would be on us to add a warning and some documentation and then mark this deprecated until v3 where we can remove the few deprecated flags we have, I don't feel persuaded towards the position you're advocating for.

> Simply not removing the flag in 2.0.x would still be breaking since we no longer honor the flag

We're again in the fuzzy area of how "breaking" this change is. Given the below:

- _the actual value of this configuration is inevitably going to match our automation_ as @shaneutt pointed out
- we're viewing this as a fix, and we're reasonably confident that there exists no scenario where the change we're making will have a negative impact on anyone's configuration (at least that's the impression I have)

I'm tempted to think that leaving the flag there, but without any effect (other than printing a deprecation warning, I suppose) can get the root issue resolved with apparent full backwards-compatibility.,

Therefore, my call would be to leave the flag there, mark it as deprecated, and print a deprecation warning when one uses it.

[the remark this answers is not in the document]
Comment (rainest) [reply CHANGELOG.md]: The flag is now deprecated instead: https://github.com/Kong/kubernetes-ingress-controller/pull/2053/commits/a71c15d5883c870536eb7b702b8e4da37dad864b

**Comment by rainest** [comment]
Rebased to roll PR changes into appropriate commits, since we don't want to squash this/will need to cherry-pick pieces into a 2.0.x backport.

**Review by shaneutt — APPROVED** [review]
